ldap/servers/slapd/log.c

```diff
@@ -2417,25 +2417,41 @@ auditfail_log_openf(char *pathname, int locked)
 static int
 vslapd_log_audit(const char *log_data)
 {
+    LogBufferInfo *lbi = loginfo.log_audit_buffer;
     char buffer[SLAPI_LOG_BUFSIZ];
-    time_t tnl;
+    time_t tnl = slapi_current_utc_time();
     int32_t blen = TBUFSIZE;
     int32_t rc = LDAP_SUCCESS;
+    int32_t msg_len = strlen(log_data);
 
 #ifdef SYSTEMTAP
     STAP_PROBE(ns-slapd, vslapd_log_audit__entry);
 #endif
 
+    /*
+     * An audit entry can be larger than the buffer, flush the current buffer
+     * and write new audit entry directly to the file
+     */
+    if (msg_len > SLAPI_LOG_BUFSIZ) {
+        PR_Lock(lbi->lock);
+        log_flush_buffer(lbi, SLAPD_AUDIT_LOG, 0 /* do not sync to disk */);
+        PR_Unlock(lbi->lock);
+        /* Now directly write the entire message to disk */
```

Comment:
Something looks fuzzy with the lock handling.
There is a potential race condition if another thread writes in the buffer  between  PR_Unlock(lbi->lock); and LOG_AUDIT_LOCK_WRITE(); 
Furthermore that code bypass all the log rotation handling which can be an issue if there is only large entries 
My felling is that you should rather fake a second bufferinfo and do everything on a single lock:
  PR_Lock(lbi->lock);
  log_flush_buffer(lbi, SLAPD_AUDIT_LOG, 0 /* do not sync to disk */);
  LogBufferInfo lbi2 = *lbi;
  lbi2.top = log_data;
  lbi2.current = log_data+msg_len;
  lbi2.maxsize = msg_len;
  log_flush_buffer(&lbi2, SLAPD_AUDIT_LOG, 0 /* do not sync to disk */);
  PR_Unlock(lbi->lock);
  return 0;

Reply:
Log rotation checks happen in log_flush_buffer() but yes the buffer can start to fill up again, but its ok since we are adding our update directly and should be in the correct order.  I could hold on to the log buffer lock through the entire process, but we have code in log_append_buffer2() that avoids holding this lock for a long time for performance reasons - so this is why I mimicked that code.

As for creating a second buffer, the update could exceed the buffer again and again.   Imagine adding 20K members to a group (this will exceed the buffer maxsize 20 times). It starts to become really complicated.   I'm just trying to keep it simple and just directly write the whole update at one shot.

Reply:
Hi agree about log rotation, I realized it just after having written the comment. :wink:

Sorry but both log_append_buffer2 and log_append_audit_buffer hold the lbi->lock buffer when effectively writing in the file (i.e when calling log_flush_buffer )
The only thing that is done out of the lock is to copy the data in the reserved area within the buffer 
 (but there is no buffer in the overlong message case)
==> you must hold lbi->lock buffer while writing directly on the file

Faking LogBufferInfo is  an overhead as log_flush_buffer will mostly calling LOG_WRITE_NOW_NO_ERR anyway
(That is why I also removed that idea)
But you are wrong about "the update could exceed the buffer again and again." :
The idea was to generate a LogBufferInfo using  directly the message as buffer (and not an external BUFSIZ limited buffer):
lbi2.top = log_data;
lbi2.current = log_data+msg_len;
lbl2.maxsize = msg_len;
lbi2.lock = NULL;
lbi2.refcount = 0;

BTW I think that a final \n must also be written in the file.